src/index.ts

```diff
@@ -935,15 +935,43 @@ export function createForgoInstance(customEnv: any) {
       renderResult.nodes.length -
       totalNodesAfterRender;
 
-    const newIndex =
+    // Since we have re-rendered, we might need to delete a bunch of nodes from the previous render.
+    // That list begins from currentIndex + num nodes in latest render.
+    // Delete up to deleteFromIndex + componentState.nodes.length - numNodesReused,
+    //  in which componentState.nodes.length is num nodes from previous render.
+    const deleteFromIndex =
       insertionOptions.currentNodeIndex + renderResult.nodes.length;
 
-    markNodesForUnloading(
+    const deletedNodes = markNodesForUnloading(
       insertionOptions.parentElement.childNodes,
-      newIndex,
-      newIndex + componentState.nodes.length - numNodesReused
+      deleteFromIndex,
+      deleteFromIndex + componentState.nodes.length - numNodesReused
     );
 
+    /*
+     * transferredState is the state that's already been remounted on a different node.
+     * Components in transferredState should not be unmounted, since this is already
+     * being tracked on a different node. Hence transferredState needs to be removed
+     * from deletedNodes.
+     */
+
+    const transferredState =
+      renderResult.nodes.length > 0 ? statesToAttach : [];
+
+    // Patch state in deletedNodes to exclude what's been already transferred.
+    for (const deletedNode of deletedNodes) {
```

Comment:
What do you think of having `markNodesForUnloading` take a callback to handle this? 

```js
markNodesFolUnloading(
  ...theUusualParams,
  (state) => {
    if (state) { ... }
  }
)
```

This saves looking up the state twice (once inside the function and again in this loop), saves building an array inside the function, and avoids having both `deletedNodes` and `justDeletedNodes`, which is kinda confusing.

It does presume that the work can be done in the middle of `markNodesForUnloading()`.

Reply:
I like your idea. In addition to what you said, it moves the mutation (of state) into `markNodesForUnloading()` - which makes it more self-contained, and obvious. Will implement.